Add Accept-Language parsing and prioritizing

tobias- opened the pull request:
Motivation:

Accept-Language is a client supplied header detailing what locales the
client user prefers and the weight of the preference. It's
mostly useful to detect the default language of the site.

Modifications:

Added method `acceptLanguages()` and `selectLocale(supportedLocales)` to
`RequestHeaderGetters`. `acceptLanguages()` is lazy and cached.

Added `acceptLanguages(acceptedLanguages)` to `RequestHeadersBuilder`.

Closes #3179

tobias- commented:
I'm not sure if the `getLocale()` method is a good idea, as it is not the same as the [RFC4647 Lookup](https://tools.ietf.org/html/rfc4647#section-3.4). Personally I would like it, because imho it's the most sane/expected version as it is the same as [obsoleted RFC2616 Accept-Language](https://tools.ietf.org/html/rfc2616#section-14.4) (the HTTP/1.1 protocol). It probably needs more documentation, to specify that it uses [RFC4647 Basic Filtering]

Reading the spec [RFC7231 Accept-Language](https://tools.ietf.org/html/rfc7231#section-5.3.5) 
the line `Implementations can offer the most appropriate matching scheme for their requirements.` seems to allow
picking and choosing your own `Locale` selection algorithm.

```
        final HttpHeaders headers = HttpHeaders.of(
                HttpHeaderNames.ACCEPT_LANGUAGE,
                "de-us ;q=0.9, *;q=0.8,en;q=0.95, en-US;q=0.98"
        );
        final List<Locale> supportedLocales = ImmutableList.of(
                Locale.GERMANY,
                Locale.UK
        );
        final List<LanguageRange> acceptLanguages = extractAcceptLanguages(headers);
        assertEquals(
                ImmutableList.of(
                        new LanguageRange("en-US", 0.98),
                        new LanguageRange("en", 0.95),
                        new LanguageRange("de-us", 0.9),
                        new LanguageRange("*", 0.8)
                ),
                acceptLanguages
        );
        // It will fail here with en_GB != null
        assertEquals(Locale.UK, Locale.lookup(acceptLanguages, supportedLocales));
```

tobias- commented:
Also, should this be documented somewhere?

trustin commented:
> Also, should this be documented somewhere? (I mean aside from the javadoc).

I guess the routing API is largely undocumented except some examples, so I guess it's fine as it is for now. 😓

trustin commented on core/src/main/java/com/linecorp/armeria/server/DefaultRoutingContext.java:
```diff
@@ -133,6 +144,17 @@ public MediaType contentType() {
+
+    @Nullable
+    @Override
+    public Locale getLocale(Collection<Locale> supportedLocales) {
```
How about `selectLocale(Iterable<Locale>)` ?

tobias- replied:
Agreed, I actually thought about this, but forgot about. I mostly put it there, because of the closeness to the media type parser.

It might even be preferable to have the `Accept` in `HttpRequest` as well, and just let `HttpRoutingContext` use that method, but that's a large change imho.

On the `selectLocale(Iterable)`, yes, much better name and signature.

trustin replied:
Alright, then let's move these methods to `HttpRequest`. `Accept` support would be even nicer, but it's up to you as always. It'd be nice if we could file an issue at least, so we can follow-up later.

tobias- replied:
Sorry, but I don't have enough time to try to figure out routing internals right now. I'm putting it in `HttpHeadersBase` if that's ok, as that's where `contentType()` parsing is done, and adding a `default` method in HttpRequest (same as for `contentType()`). I tried putting the logic in `HttpRequest`, but unless I want to put a lot of logic into an interface, that will add a lot of duplicated logic as there are about 10 classes implementing that interface.

Btw, no on the `Iterable` unfortunately. `Locale.filter()` needs a `Collection`.

ikhoon replied:
> Btw, no on the Iterable unfortunately. Locale.filter() needs a Collection.

Then, how about copy the `supportedLocales` if it is not an instance `Collection`.
For example:

```java
Locale selectLocale(Iterable<Locale> supportedLocales) {
    Collection<Locale> locales;
	if (!supportedLocales instanceof Collection) {
		locales = supportedLocales;
    } else {
		locales = ImmutableList.copyOf(supportedLocales);
    }
}

tobias- replied:
I somehow missed this comment. Going through all comments again.

trustin commented on core/src/main/java/com/linecorp/armeria/server/RoutingContext.java:
```diff
@@ -81,6 +84,23 @@
+     * the @{link HttpHeaderNames#ACCEPT_LANGUAGE} and returning the best match
+     * according to client preference.
+     * @param supportedLocales a list of locales supported by the server.
+     * @return The best matching locale or null if no locale matches.
```
I'm curious if we should return `Locale.ROOT` instead of `null`. What do you think?

tobias- replied:
If the nullability is the issue, I'd prefer `Optional<Locale>`, because I don't really see what ROOT would represent, except that nothing was found. It can't really be used to select what texts to show to the user. My completely made up statistics says that in 99.9% of the cases when this fails, the one calling the method will fall back to a specific Locale that they support and hope that the user can read it anyway.

I imagine the code looking something like
```kotlin
selectLocale(acceptedLanguages, supportedLanguages)
    ?: Locale.US
```
```java
Optional.ofNullable(selectLocale(acceptedLanguages, supportedLanguages))
    .orElse(Locale.US)
```

if the method returns Locale.ROOT, the check would be the (imho) less clear:
```java
Optional.of(selectLocale(acceptedLanguages, supportedLanguages))
    .filter((locale) -> !locale.equals(Locale.ROOT))
    .orElse(Locale.US)
```

trustin replied:
I see. Returning `null` sounds better.

trustin reviewed (commented):
I have a little doubt - is `RoutingContext` the right place for accept-language parsing and selection? It doesn't actually seem to affect routing but just provides a utility for selecting a preferred language.

Do you have a plan to make `Router` route requests to different services based on a `accept-language` header? If not, we should perhaps move this functionality somewhere else, e.g. `HttpRequest`.

tobias- commented:
Moved everything around. If you still would like to move the logic to HttpRequest, let me know, but please also tell me where to put the logic, as like I said, there are multiple implementations of that interface, and only a select few extend something that they have in common.

 Do you want `Locale.ROOT` or `Optional<>`? If so, let me know and I'll fix it. I'm still opposed to `Locale.ROOT`, but if you want that, I'll return that.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/AbstractHttpHeadersBuilder.java:
```diff
@@ -39,6 +44,18 @@ final HttpHeadersBase newSetters(HttpHeadersBase parent, boolean shallowCopy) {
     }
 
+    @Nullable
+    public final List<LanguageRange> acceptLanguages() {
```
Shouldn't we put this into `DefaultRequestHeadersBuilder` because it's a request header?

tobias- replied:
Moved to `HttpHeadersBase` as per `uri()` and `path()` and made `package`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/HttpHeaderGetters.java:
```diff
@@ -49,6 +52,27 @@
+     * in the order of client-side preferences. If the client does not send the header, this will contain only a
+     * wild card {@link LanguageRange}.
+     */
+    List<LanguageRange> acceptLanguages();
```
ditto
Shouldn't we put this into `RequestHeaderGetters`?

tobias- replied:
Done

minwoox commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -254,6 +263,33 @@ final void contentType(MediaType contentType) {
+    @Nullable
+    @Override
+    public Locale selectLocale(Collection<Locale> supportedLocales) {
+        return acceptLanguages()
```
Let's put:
`requireNonNull(supportedLocales, "supportedLocales");`

tobias- replied:
Done

minwoox commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -254,6 +263,33 @@ final void contentType(MediaType contentType) {
+    public Locale selectLocale(Collection<Locale> supportedLocales) {
+        return acceptLanguages()
+                .stream()
+                .flatMap((it) -> Locale.filter(ImmutableList.of(it), supportedLocales).stream())
```
nit: `(it)` -> `it`

tobias- replied:
fixed. I'm starting to forget my Java :\

minwoox reviewed (commented):
Thanks! Just left some nits. 😄

minwoox commented:
Also, how about adding to `AggregatedHttpRequest` with `uri()` and `contentType()`? so that `HttpRequest` and `AggregatedHttpRequest` have the same getters. /cc @trustin and @ikhoon

trustin commented:
@tobias- The new version looks better to me. Thanks! Could you address @minwoox's comments before I do another pass?

tobias- commented:
> @tobias- The new version looks better to me. Thanks! Could you address @minwoox's comments before I do another pass?

Already omw. Trying to figure out how to work with DefaultRequestHeadersBuilder and NOT having the implementation in HttpHeadersBase (which is the `CONTAINER` for DefaultRequestHeadersBuilder).

trustin commented:
> Already omw. Trying to figure out how to work with DefaultRequestHeadersBuilder and NOT having the implementation in HttpHeadersBase (which is the `CONTAINER` for DefaultRequestHeadersBuilder).

How about checking how we implemented `RequestHeaderGetters.uri()`?

tobias- commented:
> > Already omw. Trying to figure out how to work with DefaultRequestHeadersBuilder and NOT having the implementation in HttpHeadersBase (which is the `CONTAINER` for DefaultRequestHeadersBuilder).
> 
> How about checking how we implemented `RequestHeaderGetters.uri()`?

The problem is that that is also implemented in `HttpHeadersBase.java`, so not exclusively for request. I haven't found anything that is only implemented in the request-parts, but I may just not have looked hard enough.

I've moved everything around now and added it as another commit. It is broken, and includes a test case that proves that. _Please_ squash when merging as the first commit is almost completely replaced by the second, and just kept for record during the PR discussion.

tobias- commented:
> @tobias- The new version looks better to me. Thanks! Could you address @minwoox's comments before I do another pass?

Thanks! @trustin I think I addressed all the comments now.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,35 @@ public RequestHeadersBuilder authority(String authority) {
+    @Nullable
+    public Locale selectLocale(Collection<Locale> supportedLocales) {
+        final HttpHeadersBase getters = getters();
+        if (getters instanceof RequestHeaderGetters) {
```
Could do just:
```java
final HttpHeadersBase getters = getters();
if (getters != null) {
    return getters.selectLocale(supportedLocales);
}
return null;
```

minwoox commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,35 @@ public RequestHeadersBuilder authority(String authority) {
+        set(
+                HttpHeaderNames.ACCEPT_LANGUAGE,
+                acceptLanguagesValue
+        );
```
Let's merge these into one line because this is Java. 🤣  
`set(HttpHeaderNames.ACCEPT_LANGUAGE, acceptLanguagesValue);`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,35 @@ public RequestHeadersBuilder authority(String authority) {
+        return DefaultRequestHeaders.ACCEPT_LANGUAGE_WILDCARD;
+    }
+
+    public RequestHeadersBuilder acceptLanguages(List<LanguageRange> acceptLanguages) {
```
missing `@Override`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,35 @@ public RequestHeadersBuilder authority(String authority) {
+    }
+
+    @Nullable
+    public Locale selectLocale(Collection<Locale> supportedLocales) {
```
missing `@Override`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,35 @@ public RequestHeadersBuilder authority(String authority) {
+        if (getters != null) {
+            return getters.acceptLanguages();
+        }
+        return DefaultRequestHeaders.ACCEPT_LANGUAGE_WILDCARD;
```
nit: `HttpHeadersBase.ACCEPT_LANGUAGE_WILDCARD;`

tobias- replied:
Removed `DefaultRequestHeaders.ACCEPT_LANGUAGE_WILDCARD` completely and returns null.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:
```diff
@@ -318,6 +321,31 @@ default MediaType contentType() {
+
+    /**
+     * This matches the {@link Locale}s supported by the server to
+     * the @{link HttpHeaderNames#ACCEPT_LANGUAGE} and returning the best match
```
nit: `{@link ...`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:
```diff
@@ -318,6 +321,31 @@ default MediaType contentType() {
+     * the @{link HttpHeaderNames#ACCEPT_LANGUAGE} and returning the best match
+     * according to client preference. It does this via <s>Basic Filter</s>ing each
+     * {@link LanguageRange} and picking the first match. This is the "classic"
+     * algorithm described in @see
```
I think we can remove `@see`

minwoox reviewed (approved):
Just nits. 😄 
Thanks a lot!

tobias- commented on core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java:
```diff
@@ -62,4 +66,25 @@
+
+    /**
+     * This matches the {@link Locale}s supported by the server to
+     * the @{link HttpHeaderNames#ACCEPT_LANGUAGE} and returning the best match
```
{@link

tobias- commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -55,6 +62,9 @@
 
+    static final List<LanguageRange> ACCEPT_LANGUAGE_WILDCARD =
+            ImmutableList.copyOf(LanguageRange.parse("*"));
+
```
Should this be exposed somewhere?

tobias- replied:
> I cannot find any LanguageRange with "*", and the `sun.util.Locale.LocaleMatcher` that contains e.g. the actual implementation of the LanguageRange.parse() does not use a constant.
> 
> I can add that class, but it will be for just the single LanguageRange, as the problem is that a `LanguageRange` contains both the `Locale` and a weight. It might be that there are certain combinations of `Locale` and weight that are common, but I believe they would be very localized to certain countries. It may be that even USA and UK does not share any `LanguageRange`s at all.
> 
> I have no opinion on where to put that constant, so what do you want? Btw, if the `acceptLanguages()`-method is made to return `null` instead of accept all, I don't see a need for Armeria to define this constant.

Removed the `LanguageRange` constant completely as I'm returning null from acceptLanguages() instead. After thinking about it again, that was the better solution imho.

tobias- commented:
Sorry for the low quality iteration of the PR. This should be a bit better. @minwoox Thanks for all the comments.

minwoox commented:
> Sorry for the low quality iteration of the PR

No, it's not. 🤣  please don't worry about it. I appreciate your contribution and interest in our framework. 😄

minwoox commented:
https://ci.appveyor.com/project/line/armeria/builds/36400403/job/htcy74uxw34iv51x#L1356
Oops, the test failed. Could you check it please?

tobias- commented:
> https://ci.appveyor.com/project/line/armeria/builds/36400403/job/htcy74uxw34iv51x#L1356
> Oops, the test failed. Could you check it please?

It seems master is unstable as well because of the same test case?

tobias- commented:
Whoever set up testing for all the different supported JVM's, :heart: to you. Java 8 didn't have `LangaugeRange.toString()` implemented and that would've been "fun" to try to figure out later.

trustin commented on core/src/main/java/com/linecorp/armeria/common/RequestHeadersBuilder.java:
```diff
@@ -90,6 +92,13 @@ default RequestHeadersBuilder authority(Endpoint endpoint) {
+     * @param acceptedLanguages the accepted languages.
+     * @return {@code this}
+     */
+    RequestHeadersBuilder acceptLanguages(List<LanguageRange> acceptedLanguages);
```
`Iterable` could be more useful.

trustin commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -174,6 +184,32 @@ URI uri() {
+
+        final List<LanguageRange> acceptLanguages = new ArrayList<>(4);
+        for (String acceptHeader : acceptHeaders) {
+            acceptLanguages.addAll(LanguageRange.parse(acceptHeader));
```
What should we do when we see a bad header? Raise an exception or silently ignore? Other methods, such as `contentType()`, ignore bad values. When all header values are invalid, should we treat it as `ACCEPT_LANGUAGE_WILDCARD`?

tobias- replied:
Added test as well.

tobias- commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +118,31 @@ public RequestHeadersBuilder authority(String authority) {
+        if (getters != null) {
+            return getters.acceptLanguages();
+        }
+        return HttpHeadersBase.ACCEPT_LANGUAGE_WILDCARD;
```
I changed it according to your suggestion/question to return null on parse error, so while not it's possible to differ from `accept-language: *' and lack of `accept-language` header, it's not impossible to differ between no _valid_ header and no header at all. Tbh, I think that's ok, as I can't see any real value to the server side to differ between the two cases. If there is a case, it could be solved by checking for the existence of the header.

tobias- commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeaders.java:
```diff
@@ -16,16 +16,25 @@
     @Nullable
     private URI uri;
+    private final Supplier<List<LanguageRange>> acceptLanguage =
+            Suppliers.memoize(super::acceptLanguages);
```
Fixed

tobias- commented:
> Also, how about adding to `AggregatedHttpRequest` with `uri()` and `contentType()`? so that `HttpRequest` and `AggregatedHttpRequest` have the same getters. /cc @trustin and @ikhoon

@minwoox Was that for me, because I don't understand (and just skipped it when you wrote in on the grounds of the /cc)?

tobias- commented:
@minwoox Re-requested review, because too much has changed

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -174,11 +181,46 @@ URI uri() {
+    Locale selectLocale(Collection<Locale> supportedLocales) {
+        requireNonNull(supportedLocales, "supportedLocales");
+        final List<LanguageRange> languageRanges = acceptLanguages();
+        if (languageRanges == null) {
+            return null;
+        }
+        return languageRanges
+                .stream()
+                .flatMap(it -> Locale.filter(ImmutableList.of(it), supportedLocales).stream())
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Nullable
+    List<LanguageRange> acceptLanguages() {
+        final List<String> acceptHeaders = getAll(HttpHeaderNames.ACCEPT_LANGUAGE);
+        if (acceptHeaders.isEmpty()) {
+            return null;
+        }
+
+        try {
+            final List<LanguageRange> acceptLanguages = new ArrayList<>(4);
+            for (String acceptHeader : acceptHeaders) {
+                acceptLanguages.addAll(LanguageRange.parse(acceptHeader));
+            }
+            acceptLanguages.sort(comparingDouble(LanguageRange::getWeight).reversed());
+
+            return ImmutableList.copyOf(acceptLanguages);
+        } catch (IllegalArgumentException e) {
+            // If any port of any of the headers is ill-formed
+            return null;
+        }
+    }
+
     HttpMethod method() {
         final String methodStr = get(HttpHeaderNames.METHOD);
         checkState(methodStr != null, ":method header does not exist.");
         return HttpMethod.isSupported(methodStr) ? HttpMethod.valueOf(methodStr)
-                                                 : HttpMethod.UNKNOWN;
+                : HttpMethod.UNKNOWN;
```
nit: Revert?

tobias- replied:
I haven't changed it in 34b738ec4c9182305e5df22121662e133a20dc82!? (latest pushed) Something wrong with Github?

ikhoon replied:
Maybe your IDE automatically formats the part unexpectedly.
I can see the diff [here](https://github.com/line/armeria/commit/34b738ec4c9182305e5df22121662e133a20dc82#diff-115d49fc03942185a7a36423eaf6f8cb36183533cd93909744ba8a4ba4b6e402R223).
Let me fix the indent. :-)

ikhoon commented on core/src/test/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilderTest.java:
```diff
@@ -142,6 +148,120 @@ void schemeFromSessionProtocol() {
+                )
+                .build();
+        final List<LanguageRange> acceptLanguages = headers.acceptLanguages();
+        assertEquals(
+                ImmutableList.of(
+                        new LanguageRange("en-US", 0.98),
+                        new LanguageRange("en", 0.95),
+                        new LanguageRange("de-us", 0.9),
+                        new LanguageRange("*", 0.8)
+                ),
+                acceptLanguages
+        );
```
Global comment: Could you use `assertThat` of Assert4J? We prefer Assert4J when asserting in tests.
https://armeria.dev/community/developer-guide#use-assert4j-instead-of-junits-assertion-api

tobias- replied:
Oops. Fixed.

ikhoon reviewed (commented):
Looks nice! Just some minor nits. 👍

tobias- commented on core/src/main/java/com/linecorp/armeria/common/RequestHeadersBuilder.java:
```diff
@@ -90,6 +91,13 @@ default RequestHeadersBuilder authority(Endpoint endpoint) {
+     * @param acceptedLanguages the accepted languages.
+     * @return {@code this}
+     */
+    RequestHeadersBuilder acceptLanguages(Iterable<LanguageRange> acceptedLanguages);
```
I wouldn't want that, but YMMV. I have a list of languages that I support that I check against. I can ofc add another one, but imho that would be api bloat. It's not _that_ hard to do Arrays.asList()/ImmutableList.of(), but unrolling a list is not as fun.

ikhoon replied:
Agreed. We can easily add varags later if users want.

trustin replied:
How about adding now for consistency? :pray: You can use `ImmutableList.of()` or `Arrays.asList()`.

tobias- replied:
*sigh* github... When I see the comment in context it was clear.. I'll fix

tobias- commented on core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java:
```diff
@@ -62,4 +66,29 @@
+     * Returns a list of {@link LanguageRange}s that are specified in {@link HttpHeaderNames#ACCEPT_LANGUAGE}
+     * in the order of client preferences.
+     * @return All {@link LanguageRange}s of all matching headers or {@code null} if there is a parsing error
+     *     or if there is no header.
```
oh.. it's 8. I'll fix

ikhoon reviewed (approved):
Thanks! @tobias-

trustin commented on core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java:
```diff
@@ -112,4 +120,29 @@ public RequestHeadersBuilder authority(String authority) {
+                .stream(acceptLanguages)
+                .map(it -> (it.getWeight() == 1.0d) ? it.getRange() : it.getRange() + ";q=" + it.getWeight())
+                .collect(Collectors.joining(", "));
+        set(HttpHeaderNames.ACCEPT_LANGUAGE, acceptLanguagesValue);
```
What would be the sensible behavior when `acceptLanguages` is empty? :thinking:

tobias- replied:
No idea tbh.I'm inclined to say `Preconditions.checkArgument(!acceptLanguages.isEmpty, "acceptLanguages cannot be empty")`. How do you feel about that @trustin ?

trustin replied:
SGTM

trustin commented on core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:
```diff
@@ -318,6 +321,32 @@ default MediaType contentType() {
+     * @return The best matching {@link Locale} or {@code null} if no {@link Locale} matches.
+     */
+    @Nullable
+    default Locale selectLocale(Collection<Locale> supportedLocales) {
```
- I still think we should accept `Iterable<Locale>` for users. We could convert it into a `Collection` when it's not.
  - Ditto for `RequestHeaders`
- Would you mind adding `selectLocale(Locale...)` as well? I know you don't really like it but all our methods provide a varargs version :sweat_smile:

tobias- replied:
Like I said, I misinterpreted that comment as a question not a very polite order :smile:

trustin commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -174,6 +181,41 @@ URI uri() {
+    Locale selectLocale(Collection<Locale> supportedLocales) {
+        requireNonNull(supportedLocales, "supportedLocales");
+        final List<LanguageRange> languageRanges = acceptLanguages();
+        if (languageRanges == null || supportedLocales.isEmpty()) {
```
How about checking `supportedLocales` is empty before calling `acceptLanguages()`? If it's empty we don't need to parse the header.

tobias- replied:
Will do

tobias- replied:
I don't understand the following? If you mean I should add `Locale...`. If so, absolutely. It was just a cultural misunderstanding. I sincerely thought it was a question where @ikhoon wanted my opinion, not a light version of "please do it this way". :laughing: 

> How about adding now for consistency? :pray: You can use `ImmutableList.of()`.

trustin replied:
I'm not @ikhoon, so I'm not 100% sure, but I guess he thought he'll just add it later by himself since he is such a nice guy who doesn't want to waste your time doing a tedious job. :wink:

tobias- replied:
I also somehow missed his reply >_<. Sorry ikhoon.

ikhoon replied:
> I also somehow missed his reply >_<. Sorry ikhoon.

No problem. 🤣🤣🤣

trustin reviewed (approved):
A big :+1: - Thanks, @tobias-! (Please ignore the build failure. It's unrelated.)

tobias- commented:
> A big  - Thanks, @tobias-! (Please ignore the build failure. It's unrelated.)

And a thank you to all the reviewers for your patience! :bow:

minwoox reviewed (approved):
> because I don't understand (and just skipped it when you wrote in on the grounds of the /cc)?

No problem, I can handle it. 😄 
Looks good to me again, thanks a lot. 👍

trustin commented:
@tobias- Could you update the PR description to match the current version?

tobias- commented:
Oh, completely missed that. Will do

tobias- commented:
> @tobias- Could you update the PR description to match the current version?

@trustin Done. Good enough?

trustin commented:
@tobias- Looks great. I just replaced `getLocale` with `selectLocale`.
